File: ports/lua/src/unicode_lua/security/form/locale_case_inversion.lua
local casing = require("unicode_lua.security.casing")

local M = {}

local function suffix_from(input, start_idx)
  local out = {}
  for i = start_idx, #input do
    out[#out + 1] = input[i]
  end
  return out
end

local function arrays_equal(a, b)
  if #a ~= #b then
    return false
  end
  for i = 1, #a do
    if a[i] ~= b[i] then
      return false
    end
  end
  return true
end

local function first_locale_divergence(locale, input)
  local rev_prefix = {}
  for i = 1, #input do
    local cp = input[i]
    local suffix = suffix_from(input, i + 1)
    local default_lower = casing.lower_codepoint(casing.Locale.DEFAULT, rev_prefix, suffix, cp)
    local locale_lower = casing.lower_codepoint(locale, rev_prefix, suffix, cp)
    if not arrays_equal(default_lower, locale_lower) then
      return i - 1
    end
    table.insert(rev_prefix, 1, cp)
  end
  return nil
end

function M.detect(input)
  local turkish = first_locale_divergence(casing.Locale.TURKISH, input)
  if turkish ~= nil then
    return { sub = "TurkishCaseDivergence", positions = { turkish } }
  end
  local lithuanian = first_locale_divergence(casing.Locale.LITHUANIAN, input)
  if lithuanian ~= nil then
    return { sub = "LithuanianCaseDivergence", positions = { lithuanian } }
  end
  return { sub = nil, positions = {} }
end

return M

File: ports/elixir/lib/unicode_security/form/locale_case_inversion.ex
defmodule UnicodeSecurity.Form.LocaleCaseInversion do
  alias UnicodeSecurity.Casing

  def detect(input) do
    cond do
      (pos = first_locale_divergence(:turkish, input)) != nil ->
        %{sub: "TurkishCaseDivergence", positions: [pos]}

      (pos = first_locale_divergence(:lithuanian, input)) != nil ->
        %{sub: "LithuanianCaseDivergence", positions: [pos]}

      true ->
        %{sub: nil, positions: []}
    end
  end

  defp first_locale_divergence(locale, input) do
    input
    |> Enum.with_index()
    |> Enum.reduce_while([], fn {cp, i}, rev_prefix ->
      suffix = Enum.drop(input, i + 1)
      default = Casing.lower_codepoint(:default, rev_prefix, suffix, cp)
      locale_lower = Casing.lower_codepoint(locale, rev_prefix, suffix, cp)
      if default != locale_lower, do: {:halt, i}, else: {:cont, [cp | rev_prefix]}
    end)
    |> case do
      rev when is_list(rev) -> nil
      pos -> pos
    end
  end
end
